core-kotlin-modules/core-kotlin-collections-6/src/test/kotlin/com/baeldung/parallelOperationsCollections/ParallelOperationCollectionsUnitTest.kt

```diff
@@ -81,14 +77,12 @@ class ParallelOperationCollectionsUnitTest {
         val filteredPeople = people.asFlow()
             .flatMapMerge { person ->
                 flow {
-                    emit(
-                        async {
-                            person.setAdult()
-                            person
-                        }.await()
-                    )
+                    person.setAdult()
+                    Thread.sleep(1500)
+                    emit(person)
                 }
             }
+            .flowOn(Dispatchers.IO)
```

Comment:
Can you help me understand the purpose of this? I had a look at https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow-on.html

Should we put the flowOn call right after the flow call, rather than after the flatMapMerge? https://stackoverflow.com/questions/75884997/what-is-default-dispatcher-kotlin-flow-function

If not, what is the difference between those two approaches?

Reply:
Oh right, I've added it.
Because in this unit test we use _runBlocking_, so without _.flowOn(Dispatchers.IO)_, then I'm afraid flow will be run in the same context as the coroutine wrapper (in this case _runBlocking_)

Reply:
Sorry I wasn't clear. My question is, what is the difference between solutions A and B below:

A:
```
 val filteredPeople = people.asFlow()
            .flatMapMerge { person ->
                flow {
                    emit(
                        async {
                            person.setAdult()
                            person
                        }.await()
                    )
                    person.setAdult()
                    emit(person)
                }
            }
            .flowOn(Dispatchers.IO)
            .filter { it.age > 15 }.toList()
            .sortedBy { it.age }
```

B:
```
 val filteredPeople = people.asFlow()
            .flatMapMerge { person ->
                flow {
                    emit(
                        async {
                            person.setAdult()
                            person
                        }.await()
                    )
                    person.setAdult()
                    emit(person)
                }.flowOn(Dispatchers.IO)
            }
            .filter { it.age > 15 }.toList()
            .sortedBy { it.age }
            
```

Reply:
Oh, sorry, @theangrydev. I misunderstood your question.

As I understand it, code A uses the _Dispatchers.IO_ context for the entire flow, which includes _flatMapMerge_, filter, and subsequent operations. This is more efficient if all operations require execution in IO.

Code B only uses the _Dispatchers.IO_ context inside _flatMapMerge_, while _filter_ and subsequent operations will be executed in the default context or the context where the _Flow_ was originally created.

Since in unit tests we always use _runBlocking_ to run coroutines, I currently choose to use code A so that the entire flow (including filter and subsequent operations) is executed in IO.

Reply:
But now, I also remove _async {}_ inside _emit_ because I think there is no need to create new coroutines inside flow anymore.

Reply:
@hangga 

In A, can/should we put the flowOn(Dispatchers.IO) call before the flatMapMerge?

It's not very intuitive that the entire flow uses Dispatchers.IO when the flowOn call is made after the flarMapMerge

Reply:
Oh yeah, you're right.
It makes more sense to call _flowOn(Dispatchers.IO)_ before the _flatMapMerge_.
Thanks.